src/focus_manager/txm_focus_manager.js

```diff
@@ -582,169 +576,197 @@ export class TXMFocusManager {
+                    newBeyondDistance = getFarEdgeDistance(newBounds);
+                    if (newBeyondDistance <= 0) {
+                        // new focus' far edge must be beyond the current focus's far edge
+                        return;
+                    }
 
-    ensure2DArray(array) {
-        if (!array) return []; // i.e. empty, no focusables
-        if (!Array.isArray(array)) return [[array]]; // treat as a single element matrix
-        let hasSubArrays = array.find(e => Array.isArray(e));
-        if (hasSubArrays) {
-            // Already a 2D array.
-            // - ensure single top-level elements become rows
-            // - ensure shorter rows are extend their last element to the max row length.
-            let maxRowLen = array.reduce((maxLen, e) => {
-                return Math.max(maxLen, Array.isArray(e) ? e.length : 1)
-            }, 0);
-            let result = [];
-            for (let rowIndex in array) {
-                let row = array[rowIndex];
-                if (!Array.isArray(row)) row = [row];
-                else row = row.concat(); // avoid anti-aliasing
-                if (row.length < maxRowLen) {
-                    // Extend shorter rows out.
-                    let lastElmnt = row[rowIndex.length - 1];
-                    for (let i = row.length; i < maxRowLen; i++) {
-                        row.push(lastElmnt);
+                } else {
+                    newBeyondDistance = getNearEdgeDistance(newBounds);
+                    if (newBeyondDistance < 0) {
+                        return; // new focus' near edge must be adjacent or beyond the current focus's far edge
                     }
                 }
-                result.push(row);
-            }
-            return result;
 
-        } else {
-            // Interpret as 2D matrix with a single row.
-            return [array];
+                const newRange = getLaneRange(newBounds);
+                const newMatchDistance = getFocusMatchDistance(newRange);
+                if (newMatchDistance === undefined) return; // ignoring item
+
+                if (currResult === undefined // first match
+                    || newBeyondDistance < currDistanceBeyondFocus // closer to focus along lane
+                    || newBeyondDistance == currDistanceBeyondFocus && newMatchDistance < currMatchDistance) { // a better match
+                    currResult = newFocus;
+                    currDistanceBeyondFocus = newBeyondDistance;
+                    currMatchDistance = newMatchDistance;
+                }
+            });
+            return currResult;
+        }
+
+        function getBoundsOf(focusable) {
+            return focusable && focusable.element && focusable.element.getBoundingClientRect()
+              || {top: 0, left: 0, bottom: 0, right: 0, width: 0, height: 0};
+        }
+
+        function getIntersection(bounds1, bounds2) {
```

Comment:
This was a bit confusing to follow, but doing the math it makes sense.  Does this really need to return the intersection though?  I see it getting used for `equalBounds`, but that could just use the raw focus bounds of both.  Is there a weird case perhaps?

Reply:
The intersection is used to test for buttons overlapping, e.g. the choice card yes/no case
The equalBounds is to make sure it is not completely covered, as shown by one of the unit tests.